File: Medium/35SpiralTraverse.py
# def spiralTraverse(array):
#     result = []
#     startRow, endRow = 0, len(array)-1
#     startCol, endCol = 0, len(array[0])-1

#     while startRow <= endRow and startCol <= endCol:

#         for col in range(startCol, endCol+1):
#             print(array[startRow][col])
#             result.append(array[startRow][col])

#         for row in range(startRow+1, endRow+1):
#             result.append(array[row][endCol])

#         for col in reversed(range(startCol, endCol)):
#             if startRow == endRow:
#                 break
#             result.append(array[endRow][col])

#         for row in reversed(range(startRow+1, endRow)):
#             if startCol == endCol:
#                 break
#             result.append(array[row][startCol])

#         startRow += 1
#         endRow -= 1
#         startCol += 1
#         endCol -= 1

#     return result


def spiralTraverse(array):
    result = []
    spiralFill(array, 0, len(array)-1, 0, len(array[0])-1, result)
    return result


def spiralFill(array, startRow, endRow, startCol, endCol, result):
    if startRow > endRow or startCol > endCol:
        return

    for col in range(startCol, endCol+1):
        print(array[startRow][col])
        result.append(array[startRow][col])

    for row in range(startRow+1, endRow+1):
        result.append(array[row][endCol])

    for col in reversed(range(startCol, endCol)):
        if startRow == endRow:
            break
        result.append(array[endRow][col])

    for row in reversed(range(startRow+1, endRow)):
        if startCol == endCol:
            break
        result.append(array[row][startCol])
    spiralFill(array, startRow + 1, endRow - 1,
               startCol + 1, endCol - 1, result)


print(spiralTraverse([
    [1, 2, 3, 4],
    [10, 11, 12, 5],
    [9, 8, 7, 6]
]))

File: Medium/35SpiralTraverse.js
function spiralTraverse(array) {
  const result = [];
  let startRow = 0;
  let endRow = array.length - 1;
  let startCol = 0;
  let endCol = array[0].length - 1;

  while (startRow <= endRow && startCol <= endCol) {
    for (let col = startCol; col <= endCol; col++) {
      result.push(array[startRow][col]);
    }

    for (let row = startRow + 1; row <= endRow; row++) {
      result.push(array[row][endCol]);
    }

    for (let col = endCol - 1; col >= startCol; col--) {
      if (startRow === endRow) {
        break;
      }
      result.push(array[endRow][col]);
    }

    for (let row = endRow - 1; row >= startRow + 1; row--) {
      if (startCol === endCol) {
        break;
      }
      result.push(array[row][startCol]);
    }

    startRow++;
    endRow--;
    startCol++;
    endCol--;
  }

  return result;
}
console.log("====================================");
console.log(
  spiralTraverse([
    [1, 2, 3, 4],
    [12, 13, 14, 5],
    [11, 16, 15, 6],
    [10, 9, 8, 7],
  ])
);
console.log("=========================b===========");
